File: v2/src/test/java/com/api/v2/CustomerRegistrationTest.java
package com.api.v2;

import com.api.v2.common.AddressDto;
import com.api.v2.customer.dtos.CustomerRegistrationDto;
import com.api.v2.people.dtos.PersonRegistrationDto;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.web.reactive.server.WebTestClient;

import java.time.LocalDate;

@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class CustomerRegistrationTest {

	@Autowired
	private WebTestClient webTestClient;

	CustomerRegistrationDto registrationDto = new CustomerRegistrationDto(
			new PersonRegistrationDto(
					"Leo",
					"",
					"Santos",
					LocalDate.parse("2000-12-12"),
					"123456789",
					"leosantos@mail.com",
					"1234567890",
					"male"
			),
			new AddressDto(
					"CA",
					"Sacramento",
					"Downtown",
					"90012"
			)

	);

	@Test
	@Order(1)
	void testSuccessfulRegistration() {
		webTestClient
				.post()
				.uri("api/v2/customers")
				.bodyValue(registrationDto)
				.exchange()
				.expectStatus()
				.is2xxSuccessful();
	}

	@Test
	@Order(2)
	void testUnsuccessfulRegistration() {
		webTestClient
				.post()
				.uri("api/v2/customers")
				.bodyValue(registrationDto)
				.exchange()
				.expectStatus()
				.is5xxServerError();
	}

	CustomerRegistrationDto registrationDto2 = new CustomerRegistrationDto(
			new PersonRegistrationDto(
					"Leo",
					"",
					"Santos",
					LocalDate.parse("2000-12-12"),
					"123456789",
					"leosantos@mail.com",
					"1234567890",
					"male"
			),
			new AddressDto(
					"CA",
					"Sacramento",
					"Downtown",
					"90012"
			)

	);

	@Test
	@Order(3)
	void testUnsuccessfulRegistration2() {
		webTestClient
				.post()
				.uri("api/v2/customers")
				.bodyValue(registrationDto2)
				.exchange()
				.expectStatus()
				.is5xxServerError();
	}

}

File: v2/src/test/kotlin/com/api/v2/CustomerRegistrationTest.kt
package com.api.v2

import com.api.v2.people.dtos.Address
import com.api.v2.people.enums.Gender
import com.api.v2.people.PersonRegistrationDTO
import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.MethodOrderer
import org.junit.jupiter.api.Order
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestMethodOrder
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.time.LocalDate
import java.util.UUID

@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@AutoConfigureMockMvc
class CustomerRegistrationTest {

	@Autowired
	lateinit var mockMvc: MockMvc

	@Autowired
	lateinit var objectMapper: ObjectMapper

	val personDTO = PersonRegistrationDTO(
		"Leonard",
		"",
		"Smith",
		LocalDate.of(2000,12,12),
		UUID.randomUUID()
			.toString()
			.replace("-", "")
			.subSequence(0, 10)
			.toString(),
		"leosmith@mail.com",
		"1234567890",
		Gender.MALE,
		Address(
			"Downtown",
			"LA",
			"90012"
		)
	)

	@Test
	@Order(1)
	fun `should return created when successful`() {
		mockMvc.perform(
			post("/api/v2/customers")
				.content(objectMapper.writeValueAsString(personDTO))
				.contentType(MediaType.APPLICATION_JSON)
		).andExpect(status().isCreated)
	}

	@Test
	@Order(2)
	fun `should return conflict when SIN is duplicated`() {
		mockMvc.perform(
			post("/api/v2/customers")
				.content(objectMapper.writeValueAsString(personDTO))
				.contentType(MediaType.APPLICATION_JSON)
		).andExpect(status().isConflict)
	}

	val duplicatedEmailDTO = PersonRegistrationDTO(
		"Leonard",
		"",
		"Smith",
		LocalDate.of(2000,12,12),
		UUID.randomUUID()
			.toString()
			.replace("-", "")
			.subSequence(0, 10)
			.toString(),
		"leosmith@mail.com",
		"1234567890",
		Gender.MALE,
		Address(
			"Downtown",
			"LA",
			"90012"
		)
	)

	@Test
	@Order(3)
	fun `should return conflict when email is duplicated`() {
		mockMvc.perform(
			post("/api/v2/customers")
				.content(objectMapper.writeValueAsString(duplicatedEmailDTO))
				.contentType(MediaType.APPLICATION_JSON)
		).andExpect(status().isConflict)
	}
}
